grovel/grovel.lisp

```diff
@@ -723,21 +815,42 @@ string."
 ;;; written out by PROCESS-WRAPPER-FILE once everything is processed.
 (defvar *lisp-forms*)
 
-(defun generate-c-lib-file (input-file output-defaults)
-  (let ((*lisp-forms* nil)
-        (c-file (make-c-file-name output-defaults "__wrapper")))
-    (with-open-file (out c-file :direction :output :if-exists :supersede)
-      (with-open-file (in input-file :direction :input)
-        (write-string *header* out)
-        (loop for form = (read in nil nil) while form
-              do (process-wrapper-form out form))))
+(defun djb2 (string)
+  (let ((hash 5381)
+        (wrap (- (expt 2 64) 1)))
+    (loop :for c :across string :do
+       (setf hash (mod (+ (ash hash 5) hash (char-code c))
+                       wrap)))
+    hash))
+
+(defun gen-feature-hash (features)
+  (format nil "~a_~a_~x"
+          (or (operating-system) (software-type))
+          (or (architecture) (machine-type))
+          (djb2 (format nil "~{~a~}" features))))
```

Comment:
These two functions are use when generating a name for the 'feature-specific' cache directory. 
We try to balance between human readable and terse in this implementation as:
- We dont want it to be totally readable as it would use too many characters and cause problems with windows filepath length limits
- We want the user to at least be able to see whether the OS and Architecture match their machine.
My [djb2](http://www.cse.yorku.ca/~oz/hash.html) implementation is not the prettiest but works